# 1. Downloading and assembling microbial sequence data

Welcome to the **Downloading and assembling microbial sequence data** supplementary materials. This document is written as a [Jupyter](https://jupyter.org/) notebook, which means that it contains both written documentation as well as runnable code in the same document.

There are two methods you can use to run the code here: (1) either copy-paste the bash commands to a separate terminal, or (2) you can run directly in Juptyer.

## 1.1. Copy-paste commands to a separate terminal

## 1.2. Running from within Jupyter

The easiest way to run the code is to click the **Play** button in the Jupyter notebook as shown below:

![run-jupyter.png](images/run-jupyter.png)

This will run a single cell (block of code or text) and advance to the next cell.

If you, instead, which to run everything in this notebook all at once you can find the **Kerenel** option in the menu at the top and select **Restart Kerenel and Run All Cells..**, which will run everything in this notebook all at once.

![restart-kernel-jupyter.png](images/restart-kernel-jupyter.png)

# 2. Download genomes from NCBI

In this section we will go over how to download genomes from [NCBI's Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra/) by using the [SRA Tools](https://github.com/ncbi/sra-tools) command-line software provided by NCBI.

## 2.1. Install `sra-tools` using conda

This will create a new conda environment `sra-tools` and install the package `sra-tools` in this environment.

The `-y` option here means to automatically answer `yes` to any prompts for input by conda. This is important when running using Jupyter (but can be left out if you are running via the command-line).

In [3]:
conda create -y -n sra-tools sra-tools

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/sra-tools

  added / updated specs:
    - sra-tools


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0
  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0
  curl               conda-forge/linux-64::curl-7.81.0-h2574ce0_0
  hdf5               conda-forge/linux-64::hdf5-1.10.6-nompi_h6a2412b_1114
  icu                conda-forge/linux-64::icu-69.1-h9c3ff4c_0
  krb5               conda-forge/linux-64::krb5-1.19.2-hcc1bbae_3
  libcurl            conda-forge/linux-64::libcurl-7.81.0-h2574ce0_0
  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2
  libev              conda-forge/linux-64::libev-4.33-h516909a_1
  libgcc-ng      

: 1

## 2.2. Verify `sra-tools` is working

Let's verify that the tools found in the package `sra-tools` are working properly. We can do this by running one of the installed commands, `prefetch` like so:

In [4]:
conda run -n sra-tools prefetch --help


Usage: prefetch [ options ] [ accessions(s)... ]

Parameters:

  accessions(s)                    list of accessions to process


Options:

  -T|--type <file-type>            Specify file type to download. Default: sra
  -N|--min-size <size>             Minimum file size to download in KB
                                     (inclusive).
  -X|--max-size <size>             Maximum file size to download in KB
                                     (exclusive). Default: 20G
  -f|--force <no|yes|all|ALL>      Force object download - one of: no, yes,
                                     all, ALL. no [default]: skip download if
                                     the object if found and complete; yes:
                                     download it even if it is found and is
                                     complete; all: ignore lock files (stale
                                     locks or it is being downloaded by
                                     another process - use at your own

: 1

Notice that I ran by first specifying `conda run -n sra-tools [...]`. The `conda run` part lets you specify a particular environment to activate first before trying to run a tool. It's useful in cases where you are installing each tool in a separate environment (as we are doing here).

## 2.3. Prefetch genomes (`prefetch`)

The [prefetch](https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump) command that is part of the `sra-tools` package can be used to download and store sequence read data from NCBI's Sequence Read Archive. You can run it like so:

In [7]:
conda run -n sra-tools prefetch SRR3028792


2022-01-20T23:43:56 prefetch.2.11.0: 1) 'SRR3028792' is found locally
2022-01-20T23:43:56 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies



: 1

You can even pass multiple accession numbers to `prefetch` like so:

In [8]:
conda run -n sra-tools prefetch SRR3028792 SRR3028793


2022-01-20T23:44:30 prefetch.2.11.0: 1) 'SRR3028792' is found locally
2022-01-20T23:44:30 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies

2022-01-20T23:44:31 prefetch.2.11.0: 2) 'SRR3028793' is found locally
2022-01-20T23:44:31 prefetch.2.11.0: 'SRR3028793' has 0 unresolved dependencies



: 1

The `prefetch` command downloads the data from NCBI and stores on your computer in a special directory by default. You can see which directory using the `srapath` command:

In [9]:
conda run -n sra-tools srapath SRR3028792

/home/CSCScience.ca/apetkau/ncbi/public/sra/SRR3028792.sra



: 1

For how to know what identifier to use for `prefetch`, this can be found on the NCBI website.

For example for `SRR3028792`, you can find this identifier (and the associated biological sample) at https://www.ncbi.nlm.nih.gov/sra/?term=SRR3028792

![genome-sra-identifier.png](images/genome-sra-identifier.png)

## 2.4. Convert to fastq (`fasterq-dump`)

Once you've prefetched the sequence data you can convert to the **fastq** format using `fasterq-dump` (which is named based on the older `fastq-dump` tool).

In [10]:
conda run -n sra-tools fasterq-dump SRR3028792

spots read      : 824,262
reads read      : 1,648,524
reads written   : 1,648,524



: 1

This will write two `*.fastq` files do your current directory which contain the forward (`_1.fastq`) and reverse (`_2.fastq`) reads.

We can use the `ls` command to **l**i**s**t the files and show additional information (such as file size).

In [11]:
ls -lth *.fastq

-rw-r--r-- 1 apetkau grp_apetkau 398M Jan 20 17:54 SRR3028792_2.fastq
-rw-r--r-- 1 apetkau grp_apetkau 396M Jan 20 17:54 SRR3028792_1.fastq


: 1

The `396M` part shows us that `SRR3028792_1.fastq` is **396 MB**. We can compress these files to save on storage space using the `gzip` command.

In [13]:
gzip *.fastq

: 1

In [14]:
ls -lth *.fastq.gz

-rw-r--r-- 1 apetkau grp_apetkau 138M Jan 20 17:54 SRR3028792_2.fastq.gz
-rw-r--r-- 1 apetkau grp_apetkau 110M Jan 20 17:54 SRR3028792_1.fastq.gz


: 1

Now the files end in `.fastq.gz` instead of `.fastq` and are much smaller (`SRR3028792_1.fastq.gz` is **110 MB** compared to **396 MB** for the uncompressed file).

We can also look at the file contents using the `cat` (or `zcat` for files ending in `.gz`) commands along with `head`.

In [18]:
zcat SRR3028792_1.fastq.gz | head

@SRR3028792.1 1 length=90
CTTCATAATCAGGCGATAAATGCCCACCACTTTAGGCTTTCTGGCGCGGATAGCCTCCCCAATAAAATCTTTACGCGTACGCTTTGCTTC
+SRR3028792.1 1 length=90
AABC-C--CE,-,C++@,,,<,CCC,BB,:CCE,,,;CCCCC,,6+,+++8,,:,99,,8,,:,,,,<CCEE9C,@+8+8++88C,9CCC
@SRR3028792.2 2 length=271
ATACACGCCAGCGCGAATATTTATCGCTAGCATCAGGAAAACCAACGAAACAGCCCTCGTTGATTTGCCCCCTGAACAGTGGGCTTTTTTCGCGTACAGATGGCTCCGGCATTAGACCTGTGACATCCCGATACTATTCACATTCCGCAGCTCATACTTCACTTCTGCCCGTAATACGCCTTCGGACCATGCTTACGCATGACGTGTTTATTCATCAGATCATCGTCGATGGGATTCAGCGCCGGGTTAATGCCGCGCGCAATCCACGCCA
+SRR3028792.2 2 length=271
-86A<F-@F7+C+C@77C,CFF9EE8FE7+C,CE,,,,,,,;@,,8,,,,+,,9:,9,+::,,CEE,CEEDEF7,,:,,,,,,:CEFFEB=+8+++++,,:,,49A?++++8B,,,:??,:,,:,CEFF7+8++:,:A@,>,DDDD,+++88@,@,@CCC,@@B>@,>@D*@**6,5*4*6>?***>:*<,5=>,>*;5=*,,*2*<:C+<CF7A?+++3**;:*9C))7+***29F**9)7<5))8>)*9*8;5;)7)7((-57)5(4:(
@SRR3028792.3 3 length=181
TGTCGAGATGGTGGCTGAACTGCATCGGGTCTATAGCGCCCAGACTCAGGCGTATGTCCTGTTTAATGAAGAGAGTGCACTTTCGCAAGCGCTTCTGTTGCAACGTCATGGGGAAGAAGATTTTCTTGCCTTTCTCCGGGC

: 1

The command `zcat` uncompresses any `*.gz` files and then prints out the contents. Since these are large files we likely only want to look at the first few lines in the file. The command `head` will print out the first 10 lines by default (and ignore all the rest). The `|` character tells bash that we want to take the output of `zcat SRR3028792_1.fastq.gz` (which will print all the lines of a file) and forward it into the `head` command (which will print out the first 10 lines of data and ignore the rest).

The end result is we see the first 10 lines of the `SRR3028792_1.fastq.gz` file.

We can do the same thing for the reverse reads.

In [19]:
zcat SRR3028792_2.fastq.gz | head

@SRR3028792.1 1 length=300
CTCTCAAACCTTCCTCGTTACTTTTTTCTTTCCGCCTCTCTCCTCCCCCCCCCGCCTCCCCTCTTTTCCCTTTCTCTCCTGCTTCTGCCCCTCTCTCTTCTCCTCCTCTGCCTCTTCCTCCTCCCCCTTACTTCTTCCTCTCTTTCCTCCCCCTCTCCTTCCTCCTCCCCCCCTCTCCCTCCCCCCTTCCCTCCCCTCTCCCTCCCCCCCCTCCCCCCCCCTCCCTCCCTCCTCCCCCCTCTCTCTCCCCTCTCTCCCTCCCCCCTCCCCCCCCTCTCCTCCCCCCCCCCCCTCCCCCCC
+SRR3028792.1 1 length=300
--,-8,-,;:,;,;,;,:,,,,,6;;@,<@,;;,,8,8;,66;;;,88++++++678,,,,6,,6,,,6,9:?,59,:95,,:9,:,,,,9:+995:9:,:,9,4,999,,9,94,99,,8,,,++84,,,,3,8,,,,,63,,3,,33,+++++,6,,,,,,,,,,,+++*******0******))))*)+*))))))0)))))))))(((((((((((,((,((,(((((((((((((,(())))(,((((.))(((((((((((((((((((,((()((((((((,((((,,(((((
@SRR3028792.2 2 length=300
TTGCTTCGATTTCTCTCTCCCTTCACCCGGCGCTGACTCCCCTCTCCTCTTCTCTGATGCCTCCTCACTTCATCCTTCCTCCTCCTCCCCCTTCGTATTACTTTCCTCCGTCCCTTCTGCTCTCCTCCCTCTCCCTCTTTTCTCTCTTTCACATTTCTACTTCCGCCCCCCTCTTTCCCCCCACCCACCCCCCTTTTCCGCGTGCCTCTCCACTCTCGCTTTTTCCTTCCTTCTCCTGTTTCTCCCTCTCACTCTTCCCTCTTGCTTTTTTCTTTCTCTTCTACCCCTCTGCCCCTTCCC
+SRR3028792.2 2 

: 1

# 3. Quality filter files (`fastp`)

## 3.1. Install `fastp` using conda

In [20]:
conda create -y -n fastp fastp

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/fastp

  added / updated specs:
    - fastp


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  fastp              bioconda/linux-64::fastp-0.23.2-hd36eab0_1
  isa-l              conda-forge/linux-64::isa-l-2.30.0-ha770c72_4
  libdeflate         conda-forge/linux-64::libdeflate-1.9-h7f98852_0
  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11
  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate fastp
#
# To deactivate an active environment, use
#
#     $ conda deactivate



: 1

In [22]:
conda run -n fastp fastp --help

usage: fastp [options] ... 
options:
  -i, --in1                            read1 input file name (string [=])
  -o, --out1                           read1 output file name (string [=])
  -I, --in2                            read2 input file name (string [=])
  -O, --out2                           read2 output file name (string [=])
      --unpaired1                      for PE input, if read1 passed QC but read2 not, it will be written to unpaired1. Default is to discard it. (string [=])
      --unpaired2                      for PE input, if read2 passed QC but read1 not, it will be written to unpaired2. If --unpaired2 is same as --unpaired1 (default mode), both unpaired reads will be written to this same file. (string [=])
      --overlapped_out                 for each read pair, output the overlapped region if it has no any mismatched base. (string [=])
      --failed_out                     specify the file to store reads that cannot pass the filters. (string [=])
  -m, --merge  

: 1

## 3.2. Evaluate quality of sequence reads

In [26]:
conda run -n fastp fastp --in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz –-html SRR3028792.html --json SRR3028792.json

Read1 before filtering:
total reads: 824262
total bases: 176567188
Q20 bases: 171950867(97.3855%)
Q30 bases: 163432314(92.561%)

Read2 before filtering:
total reads: 824262
total bases: 177556496
Q20 bases: 153139830(86.2485%)
Q30 bases: 134061609(75.5036%)

Read1 after filtering:
total reads: 788353
total bases: 166850080
Q20 bases: 163779407(98.1596%)
Q30 bases: 156905038(94.0395%)

Read2 after filtering:
total reads: 788353
total bases: 166698912
Q20 bases: 149106629(89.4467%)
Q30 bases: 132133429(79.2647%)

Filtering result:
reads passed filter: 1576706
reads failed due to low quality: 71818
reads failed due to too many N: 0
reads failed due to too short: 0
reads with adapter trimmed: 46646
bases trimmed due to adapters: 1033979

Duplication rate: 0.665322%

Insert size peak (evaluated by paired-end reads): 35

JSON report: SRR3028792.json
HTML report: fastp.html

fastp --in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fa

: 1

# 4. Genome assembly (`skesa`)

## 4.1. Install skesa

In [27]:
conda create -y -n skesa skesa

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/skesa

  added / updated specs:
    - skesa


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  boost              conda-forge/linux-64::boost-1.70.0-py38h9de70de_1
  boost-cpp          conda-forge/linux-64::boost-cpp-1.70.0-h7b93d67_3
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4
  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0
  icu                conda-forge/linux-64::icu-67.1-he1b5a44_0
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2
  libblas            conda-forge/linux-64::libblas-3.9.0-13_linux64_openblas
  libcblas           conda-forge/linux-64::libcblas-3.9.0-13_linux64_openblas
  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5
 

: 1

## 4.2. Assemble genome

In [29]:
conda run -n skesa skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta

skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta 

Total mates: 1576706 Paired reads: 788353
Reads acquired in  14.206267s wall, 13.930000s user + 0.240000s system = 14.170000s CPU (99.7%)

Kmer len: 19
Raw kmers: 305168274 Memory needed (GB): 5.85923 Memory available (GB): 29.8726 1 cycle(s) will be performed
Distinct kmers: 172827
Kmer count in  3.040845s wall, 115.810000s user + 7.850000s system = 123.660000s CPU (4066.6%)
Uniq kmers merging in  0.100942s wall, 0.070000s user + 0.260000s system = 0.330000s CPU (326.9%)
Adapters: 0 Reads before: 1576706 Sequence before: 333548982 Reads after: 1576706 Sequence after: 333548982 Reads clipped: 0
Adapters clipped in  3.148439s wall, 115.890000s user + 8.110000s system = 124.000000s CPU (3938.5%)

Kmer len: 21
Raw kmers: 302014862 Memory needed (GB): 5.79869 Memory available (GB): 29.783 1 cycle(s) will be performed
Distinct kmers: 5886038
Kmer count in  2.713511s wall, 113.830000s user + 7.76

: 1

# 5. Evaulate assembly quality (`quast`)

## 5.1. Install `quast`

In [30]:
conda create -y -n quast quast

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/quast

  added / updated specs:
    - quast


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  blast              bioconda/linux-64::blast-2.12.0-pl5262h3289130_0
  brotli             conda-forge/linux-64::brotli-1.0.9-h7f98852_6
  brotli-bin         conda-forge/linux-64::brotli-bin-1.0.9-h7f98852_6
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4
  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0
  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0
  certifi            conda-forge/linux-64::certifi-2021.10.8-py37h89c1867_1
  circos             bioconda/noarch::circos-0.69.8-hdfd78af_1
  curl               conda-forge/linux-64::curl-7.81.0-h494985f_0
  cycler             cond

: 1

In [32]:
conda run -n quast quast --help

QUAST: Quality Assessment Tool for Genome Assemblies
Version: 5.0.2

Usage: python /home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast [options] <files_with_contigs>

Options:
-o  --output-dir  <dirname>       Directory to store all result files [default: quast_results/results_<datetime>]
-r                <filename>      Reference genome file
-g  --features [type:]<filename>  File with genomic feature coordinates in the reference (GFF, BED, NCBI or TXT)
                                  Optional 'type' can be specified for extracting only a specific feature type from GFF
-m  --min-contig  <int>           Lower threshold for contig length [default: 500]
-t  --threads     <int>           Maximum number of threads [default: 25% of CPUs]

Advanced options:
-s  --split-scaffolds                 Split assemblies by continuous fragments of N's and add such "contigs" to the comparison
-l  --labels "label, label, ..."      Names of assemblies to use in reports, comma-separated. If cont

: 1

## 5.2. Create a quality report with `quast`

In [31]:
conda run -n quast quast SRR3028792.fasta

/home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast SRR3028792.fasta

Version: 5.0.2

System information:
  OS: Linux-4.15.0-161-generic-x86_64-with-debian-buster-sid (linux_64)
  Python version: 3.7.12
  CPUs number: 56

Started: 2022-01-20 18:22:20

Logging to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/quast_results/results_2022_01_20_18_22_20/quast.log
NOTICE: Maximum number of threads is set to 14 (use --threads option to set it manually)

CWD: /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data
Main parameters: 
  MODE: default, threads: 14, minimum contig length: 500, minimum alignment length: 65, \
  ambiguity: one, threshold for extensive misassembly size: 1000

Contigs:
  Pre-processing...
  SRR3028792.fasta ==> SRR3028792

2022-01-20 18:22:20
Running Basic statistics processor...
  Contig files: 
    SRR3028792
  Calculating N50 and L50...
    SRR3028792, N50 = 298

: 1

### 5.2.1 Zip up report for download

Let's zip up the quast report for download so we can open it in our browser. You can create a `*.zip` archive from the command-line with the `zip` command.

In [ ]:
zip -r quast.zip quast_results

# 6. Where to go from here

An assembled genome file (`SRR3028792.fasta`) often forms the basis for other downstream analysis. Here are a few examples you can try on your own:

1. Try uploading `SRR3028792.fasta` to CARD/RGI (which will identify genes associated with antimicrobial resistance): https://card.mcmaster.ca/analyze/rgi